### PR TITLE
CI pipeline and related scripts for Verrazzano install/uninstall loop and various K8S versions

### PR DESCRIPTION
**Pull request by bhatpmk**
# Description

This PR includes the following changes to address VZ-2073 -

- Various shell scripts and terraform scripts required to run Verrazzano tests
- The Jenkins scripts to run Verrazzano install-uninstall in a loop using an existing cluster, uninstall test and acceptance tests using KinD clusters (versions 1.17, 1.18, 1.19 and 1.20) from develop branch. The loop of install / uninstall will be enabled later
-  The validation in platform-operator/scripts/install/1-install-istio.sh for the Kubernetes version is modified to just log the version.
- Removed Kubernetes  version 1.16.15 from the new jobs

# How has this been tested?

Provide details about how you tested this changed, if necessary
provide instructions to reproduce the testing.

- [ ] Tested locally in my own environment
- [x] Successful acceptance test run on CI server
- [ ] Other (specify)

# Checklist 

As the author of this PR, I have:

- [ ] Checked my code follows the style guidelines, ran `go fmt`
- [x] Checked that I included or updated copyright and license notices in all files that I altered
- [ ] Added or updated godoc for all public functions
- [ ] Added or updated unit tests for any new functions I added
- [ ] Added or updated integration tests if appropriate
- [ ] Added or updated acceptance tests if appropriate
- [ ] Checked that I correctly spelled trademark and product names
- [ ] Used inclusive language and neutral tone
- [x] Not included any sensitive information or hardcoded credentials

Code reviewer, please confirm this PR:

- [ ] Addressed the requirement and meets the acceptance criteria
- [ ] Does not introduce unrelated or spurious changes
- [ ] Does not introduce any unapproved dependency
- [ ] Makes sense and it easy to understand, and/or difficult areas of code are clearly documented so that they can be understood

### CODE REVIEW

File: tests/e2e/config/scripts/create_oke_cluster.sh [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+echo 'Install OKE...'
+echo 'Create cluster...'
+cd ${SCRIPT_DIR}/terraform/cluster
+./create-cluster.sh
+status_code=$?
+if [ ${status_code:-1} -eq 0 ]; then
+    echo "OKE Cluster creation request submitted."
+    cat generated/kubeconfig > ${KUBECONFIG}
+    # Adding a Service Account Authentication Token to kubeconfig
+    # https://docs.cloud.oracle.com/en-us/iaas/Content/ContEng/Tasks/contengaddingserviceaccttoken.htm
```
Comment (kminder) [inline tests/e2e/config/scripts/create_oke_cluster.sh:62]: This needs to be removed.

**Comment by kminder** [comment]
Should all those new files use only 2021 in their copyright?

File: tests/e2e/config/scripts/create_oke_cluster.sh (hunk 1 above)
Comment (bhatpmk) [reply tests/e2e/config/scripts/create_oke_cluster.sh:62]: @kminder This is a public documentation.

File: ci/looping-test/Jenkinsfile [hunk 2]
```diff
@@ -0,0 +1,476 @@
...
+            dumpCattleSystemPods()
+            dumpNginxIngressControllerLogs()
+            dumpVerrazzanoPlatformOperatorLogs()
+            dumpVerrazzanoApplicationOperatorLogs()
+            dumpOamKubernetesRuntimeLogs()
+            dumpVerrazzanoApiLogs()
+
+            script {
+                if (params.DUMP_K8S_CLUSTER_ON_SUCCESS == true) {
+                    dumpK8sCluster('verrazzano-installation-loop-dump')
```
Comment (agv426) [inline ci/looping-test/Jenkinsfile:319]: naming convention we have been using for cluster dumps is to end the name with -cluster-dump. That makes it easier for the artifact captures to be consistently finding all of them.

maybe use "verrazzano-installation-loop-cluster-dump"

Comment (bhatpmk) [reply ci/looping-test/Jenkinsfile]: Done

File: ci/looping-test/Jenkinsfile [hunk 3]
```diff
@@ -0,0 +1,476 @@
...
+
+    parameters {
+        string (name: 'VERRAZZANO_BRANCH',
+                defaultValue: 'develop',
+                description: 'Branch or tag or commit to clone and checkout, of the Verrazzano repo',
+                trim: true)
+        choice (name: 'OKE_NODE_POOL',
+                description: 'OKE node pool configuration',
+                // 1st choice is the default value
+                choices: [ "VM.Standard2.4-2", "VM.Standard.E3.Flex-8-2", "VM.Standard.E2.2" ])
```
Comment (markxnelson) [inline ci/looping-test/Jenkinsfile:37]: i think e2.2. will be too small - tests will likely fail on that size

File: ci/looping-test/JenkinsfileCreateCluster [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+                trim: true)
+        choice (description: 'OKE node pool configuration', name: 'OKE_NODE_POOL',
+            // 1st choice is the default value
+            choices: [ "VM.Standard2.4-2", "VM.Standard.E3.Flex-8-2", "VM.Standard.E2.2" ])
+        choice (description: 'OCI region to launch OKE clusters in', name: 'OKE_CLUSTER_REGION',
+            // 1st choice is the default value
+            choices: availableRegions )
+        choice (description: 'Kubernetes Version for OKE Cluster', name: 'OKE_CLUSTER_VERSION',
+                // 1st choice is the default value
+                choices: [ "v1.17.13", "v1.16.15", "v1.18.10" ])
```
Comment (markxnelson) [inline ci/looping-test/JenkinsfileCreateCluster:38]: please make 1.18 the default

File: ci/looping-test/JenkinsfileCreateCluster [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+    }
+
+    parameters {
+        string (name: 'VERRAZZANO_BRANCH',
+                defaultValue: 'develop',
+                description: 'Branch or tag or commit to clone and checkout, of the Verrazzano repo',
+                trim: true)
+        choice (description: 'OKE node pool configuration', name: 'OKE_NODE_POOL',
+            // 1st choice is the default value
+            choices: [ "VM.Standard2.4-2", "VM.Standard.E3.Flex-8-2", "VM.Standard.E2.2" ])
```
Comment (markxnelson) [inline ci/looping-test/JenkinsfileCreateCluster:32]: i think 2.2. is too small

File: ci/uninstall/Jenkinsfile [hunk 6]
```diff
@@ -0,0 +1,476 @@
...
+
+    parameters {
+        string (name: 'VERRAZZANO_BRANCH',
+                defaultValue: 'develop',
+                description: 'Branch or tag or commit to clone and checkout, of the Verrazzano repo',
+                trim: true)
+        choice (name: 'OKE_NODE_POOL',
+                description: 'OKE node pool configuration',
+                // 1st choice is the default value
+                choices: [ "VM.Standard2.4-2", "VM.Standard.E3.Flex-8-2", "VM.Standard.E2.2" ])
```
Comment (markxnelson) [inline ci/uninstall/Jenkinsfile:33]: 2.2. too small --- default to 1.18

File: platform-operator/scripts/install/1-install-istio.sh [hunk 7]
```diff
@@ -166,12 +166,12 @@ function check_kube_version {
     local major=$(echo $kubeVer | jq -r '.serverVersion.major')
     local minor=$(echo $kubeVer | jq -r '.serverVersion.minor')
     local patch=$(echo $servVer | cut -d'.' -f 3)
-    VER_ERROR_MSG="Kubernetes serverVersion $servVer must be greater than or equal to v1.16.8 and less than or equal to v1.18.*"
+    VER_ERROR_MSG="Kubernetes serverVersion $servVer must be greater than or equal to v1.16.8 and less than or equal to v1.20.*"
```
Comment (markxnelson) [inline platform-operator/scripts/install/1-install-istio.sh:169]: just remove the check please

File: tests/e2e/config/scripts/create_kind_cluster.sh [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+SCRIPT_DIR=$(cd $(dirname "$0"); pwd -P)
+CLUSTER_NAME=$1
+PLATFORM_OPERATOR_DIR=$2
+KUBECONFIG=$3
+K8S_VERSION=$4
+KIND_IMAGE=""
+
+create_kind_cluster() {
+  if [ ${K8S_VERSION} == 1.17 ]; then
+    KIND_IMAGE="v1.17.0@sha256:9512edae126da271b66b990b6fff768fbb7cd786c7d39e86bdf55906352fdf62"
```
Comment (markxnelson) [inline tests/e2e/config/scripts/create_kind_cluster.sh:16]: please double check to make sure these are the kind 0.9.0 image sha's -- doesn't look like they are

File: tests/e2e/config/scripts/create_kind_cluster.sh [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+  
+  cd ${PLATFORM_OPERATOR_DIR}/build/scripts
+  ./cleanup.sh ${CLUSTER_NAME}
+
+  cd ${SCRIPT_DIR}/
+  echo "KinD Image : ${KIND_IMAGE}"
+  sed -i "s/KIND_IMAGE/${KIND_IMAGE}/g" kind-config.yaml
+  HTTP_PROXY="" HTTPS_PROXY="" http_proxy="" https_proxy="" time kind create cluster -v 1 --name ${CLUSTER_NAME} --wait 5m --config=kind-config.yaml
+  kubectl config set-context kind-${CLUSTER_NAME}
+  sed -i -e "s|127.0.0.1.*|`docker inspect ${CLUSTER_NAME}-control-plane | jq '.[].NetworkSettings.IPAddress' | sed 's/"//g'`:6443|g" ${KUBECONFIG}
```
Comment (markxnelson) [inline tests/e2e/config/scripts/create_kind_cluster.sh:36]: see https://github.com/verrazzano/verrazzano/pull/651/files#diff-f997a7e4bf98718d3b14cc8a3cf958da70b1a9900bd49765c49d4a491bd2af08R234
will need to update this to match

File: tests/e2e/config/scripts/delete_oke_cluster.sh [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+
+# clean up PVC
+kubectl delete pvc --all -n keycloak --timeout 2m
+# wait until OKE cleans up the deleted resources
+timeout 2m bash -c 'until kubectl get pv -A 2>&1 | grep "No resources found"; do sleep 10; done'
+
+# log what still exists, just in case
+kubectl get pvc,pv,svc -A
+
+# delete the OKE cluster
```
Comment (markxnelson) [inline tests/e2e/config/scripts/delete_oke_cluster.sh:49]: need to delete the VCN too

File: tests/e2e/verify-install/kubernetes/kubernetes_test.go [hunk 11]
```diff
@@ -61,7 +61,7 @@ var _ = ginkgo.Describe("Kubernetes Cluster",
 			func() {
 				clientset := pkg.GetKubernetesClientset()
 				versionInfo, _ := clientset.ServerVersion()
-				gomega.Expect(versionInfo.GitVersion).Should(gomega.MatchRegexp(`(v1\.1[5-8]\.*)`))
+				gomega.Expect(versionInfo.GitVersion).Should(gomega.MatchRegexp(`(v1\.1[5-9]\.*)`))
```
Comment (markxnelson) [inline tests/e2e/verify-install/kubernetes/kubernetes_test.go:64]: i'm fine to remove this check

File: tests/e2e/config/scripts/delete_oke_cluster.sh (hunk 10 above)
Comment (bhatpmk) [reply tests/e2e/config/scripts/delete_oke_cluster.sh:49]: The terraform script takes care of deleting the VCN as well.

File: ci/kind/Jenkinsfile [hunk 12]
```diff
@@ -0,0 +1,414 @@
...
+
+    parameters {
+        string (name: 'VERRAZZANO_BRANCH',
+                        defaultValue: 'develop',
+                        description: 'Branch or tag or commit to clone and checkout, of the Verrazzano repo',
+                        trim: true)
+        choice (name: 'KUBERNETES_CLUSTER_VERSION',
+                description: 'Kubernetes Version for KinD Cluster',
+                // 1st choice is the default value
+                choices: [ "1.17", "1.18", "1.19", "1.20" ])
```
[the remark this answers is not in the document]
Comment (bhatpmk) [reply ci/kind/Jenkinsfile]: Done

File: ci/looping-test/JenkinsfileCreateCluster (hunk 4 above)
Comment (bhatpmk) [reply ci/looping-test/JenkinsfileCreateCluster]: Done

File: ci/uninstall/Jenkinsfile (hunk 6 above)
Comment (bhatpmk) [reply ci/uninstall/Jenkinsfile]: I have added the support for VM.Standard.E2.4 and removed VM.Standard.E2.2. We need more validation with other shapes.

File: tests/e2e/config/scripts/create_kind_cluster.sh (hunk 8 above)
Comment (bhatpmk) [reply tests/e2e/config/scripts/create_kind_cluster.sh]: I have added the images for kind 0.9.0, except for 1.20 (which is from 0.10.0)

File: tests/e2e/config/scripts/create_kind_cluster.sh (hunk 9 above)
Comment (bhatpmk) [reply tests/e2e/config/scripts/create_kind_cluster.sh]: Done. Actually we can make use of the Makefile target from the jenkins pipeline, going forward.

File: tests/e2e/verify-install/kubernetes/kubernetes_test.go (hunk 11 above)
Comment (bhatpmk) [reply tests/e2e/verify-install/kubernetes/kubernetes_test.go]: Done

File: ci/looping-test/Jenkinsfile (hunk 3 above)
Comment (bhatpmk) [reply ci/looping-test/Jenkinsfile]: Added the support for VM.Standard.E2.4, need more testing for other shapes.

File: ci/looping-test/JenkinsfileCreateCluster (hunk 5 above)
Comment (bhatpmk) [reply ci/looping-test/JenkinsfileCreateCluster]: Added the support for VM.Standard.E2.4, need more testing for other shapes.

File: platform-operator/scripts/install/1-install-istio.sh (hunk 7 above)
Comment (bhatpmk) [reply platform-operator/scripts/install/1-install-istio.sh]: Done

**Review by agv426 — APPROVED** [review]

File: ci/looping-test/Jenkinsfile [hunk 13]
```diff
@@ -34,15 +34,15 @@ pipeline {
...
+                choices: [ "VM.Standard2.4-2", "VM.Standard.E3.Flex-8-2", "VM.Standard.E2.4" ])
         string (name: 'VERRAZZANO_OPERATOR_IMAGE',
                 defaultValue: 'NONE',
                 description: 'Verrazzano platform operator image name (in ghcr.io repo).  If not specified, the operator.yaml from VZ repo will be leveraged to create VZ platform operator',
                 trim: true)
         choice (name: 'OKE_CLUSTER_VERSION',
                 description: 'Kubernetes Version for OKE Cluster',
                 // 1st choice is the default value
-                choices: [ "v1.17.13", "v1.16.15", "v1.18.10" ])
+                choices: [ "v1.18.10", "v1.17.13", "v1.16.15"])
```
[the remark this answers is not in the document]
Comment (bhatpmk) [reply ci/looping-test/Jenkinsfile]: Done

**Review by markxnelson — APPROVED** [review]

**Pull request merged**